TracerSeq
=========

This pipeline facilitates processing of scRNA-seq TracerSeq FASTQ files to produce a TracerClones x Cells counts table.  

These functions require that FASTQ files have already been demultiplexed by sample (e.g. steps 1-3 of the inDrops.py pipeline).

This code was written in Matlab (2017a) and requires the 'Statistics and Machine Learning Toolbox'.

TracerSeq was developed and used to analyze zebrafish embryonic development in:  
**Single-cell mapping of gene expression landscapes and lineage in the zebrafish embryo.**  Wagner DE, Weinreb C, Collins ZM, Briggs JA, Megason SG, Klein AM. Science 26 Apr 2018. [doi:10.1126/science.aar4362](http://science.sciencemag.org/content/early/2018/04/25/science.aar4362)


## Usage ##

### Inputs ###

### Settings ###

### Parse TracerSeq FASTQ Files ###











